code/modules/security levels/keycard authentication.dm

```diff
@@ -53,93 +44,59 @@
 	else
 		stat |= NOPOWER
 
-/obj/machinery/keycard_auth/attack_hand(mob/user)
-	if(user.stat || stat & (NOPOWER|BROKEN))
-		user << "<span class='warning'>This device is not powered!</span>"
-		return
-	if(busy)
-		user << "<span class='warning'>This device is busy!</span>"
-		return
-
-	user.set_machine(src)
-
-	var/dat = "<h1>Keycard Authentication Device</h1>"
-
-	dat += "This device is used to trigger some high security events. It requires the simultaneous swipe of two high-level ID cards."
-	dat += "<br><hr><br>"
-
-	if(screen == 1)
-		dat += "Select an event to trigger:<ul>"
-		dat += "<li><A href='?src=\ref[src];triggerevent=Red alert'>Red alert</A></li>"
-		dat += "<li><A href='?src=\ref[src];triggerevent=Emergency Maintenance Access'>Emergency Maintenance Access</A></li>"
-		dat += "</ul>"
-		user << browse(dat, "window=keycard_auth;size=500x250")
-	if(screen == 2)
-		dat += "Please swipe your card to authorize the following event: <b>[event]</b>"
-		dat += "<p><A href='?src=\ref[src];reset=1'>Back</A>"
-		user << browse(dat, "window=keycard_auth;size=500x250")
-	return
+/obj/machinery/keycard_auth/get_ui_data()
+	var/list/data = list()
+	data["broken"] = (stat & (NOPOWER|BROKEN)) ? 1 : 0
+	data["waiting"] = waiting
+	data["authRequired"] = event_source ? 1 : 0
+	data["redAlert"] = (seclevel2num(get_security_level()) == SEC_LEVEL_RED) ? 1 : 0
+	data["emergencyMaint"] = emergency_access
+	return data
+
+/obj/machinery/keycard_auth/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, \
+					datum/tgui/master_ui = null, datum/ui_state/state = physical_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "keycard_auth", name, 350, 150, master_ui, state)
+		ui.open()
 
-
-/obj/machinery/keycard_auth/Topic(href, href_list)
-	if(..())
-		return
-	if(busy)
-		usr << "This device is busy."
-		return
-	if(href_list["triggerevent"])
-		event = href_list["triggerevent"]
-		screen = 2
-	if(href_list["reset"])
-		reset()
-
-	updateUsrDialog()
-	return
-
-/obj/machinery/keycard_auth/proc/reset()
-	active = 0
-	event = ""
-	screen = 1
-	confirmed = 0
-	event_source = null
-	icon_state = "auth_off"
-	event_triggered_by = null
-	event_confirmed_by = null
-
-/obj/machinery/keycard_auth/proc/broadcast_request()
-	icon_state = "auth_on"
-	for(var/obj/machinery/keycard_auth/KA in machines)
-		if(KA == src) continue
-		KA.reset()
-		spawn()
-			KA.receive_request(src)
-
-	sleep(confirm_delay)
-	if(confirmed)
-		confirmed = 0
-		trigger_event(event)
-		log_game("[key_name(event_triggered_by)] triggered and [key_name(event_confirmed_by)] confirmed event [event]")
-		message_admins("[key_name(event_triggered_by)] triggered and [key_name(event_confirmed_by)] confirmed event [event]")
-	reset()
-
-/obj/machinery/keycard_auth/proc/receive_request(obj/machinery/keycard_auth/source)
+/obj/machinery/keycard_auth/attack_hand(mob/user)

```

Comment:
Remove any attack_ and interact() procs, the parent machinery procs will call ui_interact for you.